rfc/rfc016-lamport-clock.md

```diff
@@ -0,0 +1,63 @@
+# RFC016 Transaction Lamport clock
+
+|  |  |
+| :--- | :--- |
+| Nuts foundation | W.M. Slakhorst |
+| Request for Comments: 016 | Nedap |
+| Updates: RFC004 | November 2021 |
+
+## Transaction Lamport clock
+
+### Abstract
+
+This RFC updates [RFC004](rfc004-verifiable-transactional-graph.md). It specifies a version two transaction format and adds additional requirements to version 1.
+All additional requirements are backwards compatible. This RFC introduces the concept of the [Lamport clock](https://en.wikipedia.org/wiki/Lamport_timestamp) to the transaction contents.
+
+### Status of document
+
+This document is currently a draft.
+
+### Copyright Notice
+
+![](../.gitbook/assets/license.png)
+
+This document is released under the [Attribution-ShareAlike 4.0 International \(CC BY-SA 4.0\) license](https://creativecommons.org/licenses/by-sa/4.0/).
+
+## 1.  Introduction
+
+Version one of the transaction format introduced an ordening based on predecessors. This creates a strict ordering but does not support the exchange of missing transactions between nodes.
+More efficient set reconciliation methods can be used when transactions can be grouped into ranges. 
+
+## 2. Terminology
+
+* **Lamport clock**: Logical clock creating a casual ordering between transactions.
+
+Other terminology comes from the [Nuts Start Architecture](rfc001-nuts-start-architecture.md#nuts-start-architecture).
+
+## 3. New transaction format
+
+Version 2 transactions remain encoded as JWS: [RFC004 ref](rfc004-verifiable-transactional-graph.md#3-transaction-format).
+
+Changes for the version 2 transactions:
+
+* The **ver** protected header MUST be 2.
+* The **lc** protected header is added. It MUST be a positive number constructed as follows:
+  * if the transaction has no entries in **prevs**, it has a lc value that is equal to **sigt**.
+  * otherwise, the value is equal to `max(prev1, ... prevN)+1`.
```

Comment:
How do we know it's equal to `max(prev1, ... prevN)+1`? The Lamport timestamp could also be a different value as long as it's more than `max(prev1, ... prevN)`.

Reply:
We will make this a condition with MUST

Reply:
I'll add a ref to RFC004 and describe the combined effect of using prevs in the TX and LC value